### PR TITLE
Popups

### PR DESCRIPTION
**Pull request by Avraham-Sason**

### CODE REVIEW

File: src/components/popup/Comps.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// utility function to compare props
+const areEqual = (prevProps: TObject<any>, nextProps: TObject<any>) => {
+    const { position: prevPosition, ...prevOtherProps } = prevProps;
+    const { position: nextPosition, ...nextOtherProps } = nextProps;
+    return isEqual(prevOtherProps, nextOtherProps);
+};
+
+// memoize the component
+const Warper = memo<PopupWrapperProps>(PopupWrapper, areEqual);
+Warper.displayName = "PopupWarper";
```
Comment (akeylesspro) [inline src/components/popup/Comps.tsx:25]: typo: PopupWrapper

File: src/components/popup/Comps.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            }}
+        >
+            <div className="bg-gray-500 text-white h-full flex items-center justify-between  hover:bg-gray-700  rounded-t-md">
+                <div className="h-full flex justify-start items-center">
+                    {!close?.noClose && (
+                        <button title={t("close")} onClick={exitPopUp} className="center text-white w-8 h-full hover:bg-[#d90d0d]  rounded-tl-md">
+                            <i className="fa-solid fa-x"></i>
+                        </button>
+                    )}
+                    <button title={t("restore")}  onClick={() => restorePopup(id)} className="center text-white w-8 h-full hover:bg-gray-500">
```
Comment (akeylesspro) [inline src/components/popup/Comps.tsx:53]: restore should appear only if popup is already minimized

File: src/components/popup/Comps.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                        <i className="fa-light fa-square"></i>
+                    </button>
+                </div>
+                <div title={minimize?.iconTitle} className="px-1 _center">
```
Comment (akeylesspro) [inline src/components/popup/Comps.tsx:57]: minimize should appear only if this option is enabled

File: src/components/popup/Comps.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                    {minimize?.icon || type}
```
Comment (akeylesspro) [inline src/components/popup/Comps.tsx:58]: why need "type" here

File: src/components/popup/types.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    zIndex?: number;
+    dragAndDrop?: boolean;
+    resize?: boolean;
+    singleton?: boolean;
+    errorMsg?: string;
+    minimize?: {
+        enable?: boolean;
+        isMinimized?: boolean;
+        icon?: React.ReactNode;
+        iconTitle?: string;
```
Comment (akeylesspro) [inline src/components/popup/types.ts:23]: what do you think to use it for ?

File: src/components/popup/types.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    headerBackground?: string;
+    zIndex?: number;
+    dragAndDrop?: boolean;
+    resize?: boolean;
+    singleton?: boolean;
+    errorMsg?: string;
+    minimize?: {
+        enable?: boolean;
+        isMinimized?: boolean;
+        icon?: React.ReactNode;
```
Comment (akeylesspro) [inline src/components/popup/types.ts:22]: what you think it may be ?

File: src/components/popup/index.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
+import React, { memo, useCallback } from "react";
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:1]: it's weird to see a lot of code in index

File: src/components/popup/types.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        noClose?: boolean;
+        onClose?: () => void | Promise<void>;
+    };
+    top?: string;
+    left?: string;
+    bottom?: string;
+    right?: string;
+    headerBackground?: string;
+    zIndex?: number;
+    dragAndDrop?: boolean;
```
Comment (akeylesspro) [inline src/components/popup/types.ts:15]: suggest to rename it to more suitable "move". dragAndDrop is something else: it's when you drag an object over another object and leave it into it, then this 2-nd object accepts the first object

File: src/components/popup/index.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        type,
+        right = "465px",
+        left = "auto",
+        top = "105px",
+        bottom,
+        headerBackground = "linear-gradient(180deg, #7D7D7D 0%, #495359 73.44%, #364046 100%)",
+        zIndex = 10,
+        dragAndDrop = true,
+        close,
+        resize,
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:23]: suggest to use default = false

File: src/components/popup/index.tsx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+import { PopupsStore } from "@/lib/store";
+import { useDragAndDrop, useResize } from "./hooks";
+import { PopUpProps } from "./types";
+
+const Popup = memo(
+    ({
+        id,
+        element,
+        type,
+        right = "465px",
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:15]: right and top will be caclulated automatically according to already displayed popups. The goal it not to overlap existing popups. We can implement it later (or now, it should not be complicated)

File: src/components/popup/index.tsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        left = "auto",
+        top = "105px",
+        bottom,
+        headerBackground = "linear-gradient(180deg, #7D7D7D 0%, #495359 73.44%, #364046 100%)",
+        zIndex = 10,
+        dragAndDrop = true,
+        close,
+        resize,
+        errorMsg,
+        maximize,
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:25]: false

File: src/components/popup/index.tsx [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        minimize,
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:26]: false

File: src/components/popup/index.tsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+        const { isDragging, startDragging, position, setPosition } = useDragAndDrop({
+            initialPosition: { top, left, right, bottom },
+        });
+
+        const { size, setSize, startResizing } = useResize({
+            resize,
+            initialWidth: 300,
+            minWidth: 300,
+            initialHeight: 150,
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:43]: I think we can do popup be auto size. Height for sure. But still the width should have some reasonable minimum

File: src/components/popup/index.tsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            minHeight: 150,
+            setPosition,
+            position,
+        });
+
+        const exitPopUp = useCallback(() => {
+            deletePopup(id);
+            close?.onClose?.();
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:51]: what about onClose to be able to return False? Then a) need to execute it before deletePopup and following according to result of onClose. If possible - check the type of onClose: if it's void - then simple execute it, if it's <Boolean> then check the result

File: src/components/popup/index.tsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+                    id={id}
+                    type={type}
+                    close={close}
+                    minimize={minimize}
+                    zIndex={zIndex}
+                    minimizedPopups={minimizedPopups}
+                    exitPopUp={exitPopUp}
+                />
+            );
+        }
```
Comment (akeylesspro) [inline src/components/popup/index.tsx:70]: I am missing the restore functionality - please explain where is it

File: src/components/popup/types.ts [hunk 16]
```diff
@@ -0,0 +1,62 @@
+export interface PopUpProps {
+    id: string;
+    element: JSX.Element;
+    type: "info" | "error" | "warning" | "interaction";
```
Comment (akeylesspro) [inline src/components/popup/types.ts:4]: do not see how we use this type. I think type should affect on default border color of the popup and header color

File: src/lib/store/PopupsStore.tsx [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+                    : popup
+            ),
+            maxZIndex: newZIndex,
+        }));
+    },
+    minimize: (id) => {
+        set((state) => {
+            const popupIndex = state.popups.findIndex((popup) => popup.id === id);
+
+            if (popupIndex !== -1) {
```
Comment (akeylesspro) [inline src/lib/store/PopupsStore.tsx:56]: how it gets -1 ?

File: src/components/popup/types.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    bottom?: string;
+    right?: string;
+    headerBackground?: string;
+    zIndex?: number;
+    dragAndDrop?: boolean;
+    resize?: boolean;
+    singleton?: boolean;
+    errorMsg?: string;
+    minimize?: {
+        enable?: boolean;
```
[the remark this answers is not in the document]
Comment (Avraham-Sason) [reply src/components/popup/types.ts:20]: done

File: src/components/popup/Comps.tsx (hunk 2 above)
Comment (Avraham-Sason) [reply src/components/popup/Comps.tsx:53]: done

File: src/components/popup/Comps.tsx (hunk 3 above)
Comment (Avraham-Sason) [reply src/components/popup/Comps.tsx:57]: done

File: src/components/popup/types.ts (hunk 16 above)
Comment (Avraham-Sason) [reply src/components/popup/types.ts:4]: change border color according to type

File: src/components/popup/components.tsx [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+ResizeHandle.displayName = "resizeHandle";
+
+// popup manager component
+const PopupManager = () => {
+    const popups = PopupsStore.popups();
+    const addPopup = PopupsStore.addPopup();
+    const managerRef = useRef<HTMLDivElement>(null);
+
+    useEffect(() => {
+        addPopup({
```
Comment (akeylesspro) [inline src/components/popup/components.tsx:91]: Please move all examples to readme.txt

File: src/components/popup/hooks.tsx [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+        (e: React.MouseEvent<HTMLDivElement>) => {
+            e.preventDefault();
+            if (popupRef.current) {
+                const popupRect = popupRef.current.getBoundingClientRect();
+                setIsDragging(true);
+                setOffset({
+                    x: e.clientX - popupRect.left,
+                    y: e.clientY - popupRect.top,
+                });
+                document.body.classList.add("select-none");
```
Comment (akeylesspro) [inline src/components/popup/hooks.tsx:19]: please fix it

File: src/components/popup/types.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+        isMinimized?: boolean;
+    };
+    headerIcon?: React.ReactNode;
+    headerTitle?: string;
+    maximize?: {
+        enabled?: boolean;
+        height?: number;
+        width?: number;
+    };
+
```
Comment (akeylesspro) [inline src/components/popup/types.ts:30]: remove line

File: src/app/login/Login.tsx [hunk 22]
```diff
@@ -45,7 +45,7 @@ const Login = () => {
 
     return (
         <div className="_full _center px-3">
-            <Image className="fixed top-4 left-4" width={214} height={57} src="/images/akeyless_logo_big.png" alt="akyeless_logo" />
+            <Image className="fixed top-4 left-4" width={214} height={57} src="/images/akeyless_logo_big.png" alt="akyeless_logo" fetchPriority="high" />
```
[the remark this answers is not in the document]
Comment (Avraham-Sason) [reply src/app/login/Login.tsx]: done

**Pull request merged**